### ISSUE TITLE
Host: randomness tests for the stored data

### ISSUE THREAD

**Issue by starius** [NebulousLabs/Sia#2260]
Attackers can upload non-encrypted illegal files to hosts, which can be dangerous for host operators from legal perspective. (The uploaded data can be forbidden to be stored in the country.) Even if the host operator is clean as a whistle, law enforcement people often lack education in this area and can still seize the hardware and arrest the operator (this [happened](http://web.archive.org/web/20141004142101/http://raided4tor.cryto.net/) with Tor operators).

Hosts can reduce this risk by doing randomness tests of the uploaded data and reject if it seems to be non-encrypted. We can use the fact that the encrypted data can't be distinguished from random data reliably. As a starting point we can check that `net/http.DetectContentType` returns `"application/octet-stream"` for the data. Also we can measure Shannon Entropy and reject data if the entropy value is low.

**Comment by lukechampine** [comment]
Here's a simple function that returns the randomness (as a value between 0 and 1) of a given `[]byte`:
```go
func ShannonEntropy(data []byte) float64 {
	frq := make(map[byte]int)
	for _, i := range data {
		frq[i]++
	}

	var sum float64
	for _, v := range frq {
		f := float64(v) / float64(len(data))
		sum -= f * math.Log2(f)
	}

	return sum / 8
}
```
It returns 0 when `data` is a single repeated byte, and 0.99999 when `data` 4MB of `crypto/rand`. So 0.999 seems like a reasonable cut-off. If the cut-off is too low, a malicious renter could store a bit of plaintext alongside a lot of randomness. In that case, maybe a "sliding window" approach makes more sense?

**Comment by starius** [comment]
I tried this function with a video file of format:
RIFF (little-endian) data, AVI, 720 x 400, 23.98 fps, video: XviD, audio: MPEG-1 Layer 3 (stereo, 48000 Hz)
In the middle of a file, the function returned 0.99918 for 1 megabyte of data.
Also, `DetectContentType` returned `"application/octet-stream"`.

**Comment by lukechampine** [comment]
Hmm. I guess in practice it will be difficult (theoretically impossible) to distinguish between data that's encrypted and data that's compressed. A sliding window still might work, since even a compressed format will have a low-entropy header or footer.

**Comment by lukechampine** [comment]
won't a zlib of a zlib-compressed (but unencrypted) file be larger than the original?
maybe we could check for the magic header bytes for a number of formats. But then you run the risk of false positives...

**Comment by petabytestorage** [comment]
Yes compressed files are indistinguishable from random files if you only look at empirical entropy in most cases. But the easiest way to detect compressed files is to try to decompress them based on the magic numbers in the first few bytes of the file or just try a list of popular decompressors. And same with de-archiving. An encrypted file will never accidentally match a zlib compressed file because the header and checksums will not pass when zlib tries to decompress.

**Comment by starius** [comment]
Faster function for Shannon entropy:

```go
func ShannonEntropy0(data []byte) float64 {
        var frq [256]int
        for _, v := range data {
                frq[v]++
        }
        var sum float64
        den := float64(len(data))
        for _, v := range frq {
                if v > 0 {
                        f := float64(v) / den
                        sum -= f * math.Log2(f)
                }
        }
        return sum / 8
}
```

I measured Shannon entropy of 4MiB blocks from `/dev/urandom`, video files and gzip compressed files.

In sample of 1000 random blocks, minimum entropy was 0.9999929
In AVI files maximum entropy was 0.9654276
In WMV files: 0.9994110
In MP4 files: in one it was 0.9992273, but in other 0.9999933 (in that file 2 of 3700 blocks have entrory >= 0.9999929).
Other mp4 file has max entropy 0.9999770. After I compressed it with gzip, it became 0.9999940 (in that file 15 of 1200 blocks have entrory >= 0.9999929).

If we hosts reject sectors with entrory < 0.9999929, almost all raw video and most of gzip compressed data will be rejected and only ~ 1/1000 of random data will be rejected.

We can move the threshold to **0.999992** to further decrease probability of false positives. In my test of 5000 random sectors none of them had entropy lower than 0.999992 (the minimum was 0.9999926). The mp4 video file had about 10 of 3700 sectors with entropy > 0.999992. In the compressed file it was 50 of 1200 sectors.

Renters getting the error "too low shannon entropy" can just change the encryption key. A video file requires much more work to increase its entropy.
